Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -177,6 +177,199 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
     return false;
 }
 
+// Generate positions to be queried based on the sampler type.
+void TerrainSystem::GenerateQueryPositions(const AZStd::span<AZ::Vector3>& inPositions,
+    AZStd::vector<AZ::Vector3>& outPositions,
+    Sampler sampler, size_t indexStepSize) const
+{
+    const float minHeight = m_currentSettings.m_worldBounds.GetMin().GetZ();
+    for (size_t i = 0, iteratorIndex = 0; i < inPositions.size(); i++, iteratorIndex += indexStepSize)
+    {
+        switch(sampler)
+        {
+        case AzFramework::Terrain::TerrainDataRequests::Sampler::BILINEAR:
+            {
+                AZ::Vector2 normalizedDelta;
+                AZ::Vector2 pos0;
+                ClampPosition(inPositions[i].GetX(), inPositions[i].GetY(), pos0, normalizedDelta);
+                const AZ::Vector2 pos1(pos0.GetX() + m_currentSettings.m_heightQueryResolution,
+                    pos0.GetY() + m_currentSettings.m_heightQueryResolution);
+                outPositions[iteratorIndex].Set(pos0.GetX(), pos0.GetY(), minHeight);
+                outPositions[iteratorIndex + 1].Set(pos1.GetX(), pos0.GetY(), minHeight);
+                outPositions[iteratorIndex + 2].Set(pos0.GetX(), pos1.GetY(), minHeight);
+                outPositions[iteratorIndex + 3].Set(pos1.GetX(), pos1.GetY(), minHeight);
+            }
+            break;
+        case AzFramework::Terrain::TerrainDataRequests::Sampler::CLAMP:
+            {
+                AZ::Vector2 normalizedDelta;
+                AZ::Vector2 clampedPosition;
+                ClampPosition(inPositions[i].GetX(), inPositions[i].GetY(), clampedPosition, normalizedDelta);
+                outPositions[iteratorIndex].Set(clampedPosition.GetX(), clampedPosition.GetY(), minHeight);
+            }
+            break;
+        case AzFramework::Terrain::TerrainDataRequests::Sampler::EXACT:
+            [[fallthrough]];
+        default:
+            outPositions[iteratorIndex].Set(inPositions[i].GetX(), inPositions[i].GetY(), minHeight);
+            break;
+        }
+    }
+}
+
+void TerrainSystem::GenerateInputPositionsFromRegion(
+    const AZ::Aabb& inRegion,
+    const AZ::Vector2& stepSize,
+    AZStd::span<AZ::Vector3> inPositions) const
```

Comment:
Suggestion: Consider making this `AZStd::vector<AZ::Vector3> GenerateInputPositionsFromRegion(const AZ::Aabb& inRegion, const AZ::Vector2& stepSize)` where it creates the vector, instead of operating on it as a span<>. With the way it's written now, the inPositions vector will have to get fully filled out with default values before getting set with new ones. If you create the vector here like this:
```
AZStd::vector<AZ::Vector3> inPositions;
inPositions.reserve(numSamplesX * numSamplesY);
```
you can allocate the memory without having to construct the Vector3s, and then just use emplace_back() to push the constructed Vector3s.
Usage of the method would then look like:
```
AZStd::vector<Vector3> inPositions = GenerateInputPositionsFromRegion(inRegion, stepSize);
```

Reply:
You can solve the filling with default values problem by doing a `resize_no_construct()`, but still agree with Mike's suggestion for a more natural api. However, it'd be OK to have both. You could keep this function and make another trivial function that creates / returns a vector then uses the version that takes a span internally.

Reply:
Updated to use Mike's suggestion